File: backend/src/summarizer.rs
pub async fn summarize_content(
    url: &String,
    content: &String,
    model: &String,
) -> Result<(String, u64), Box<dyn std::error::Error>> {
    let client = reqwest::Client::new();
    let prompt: String = String::from(format!(
        "Following is the text extracted from the url {} and i want you to summarize it and give me your response in a json format with the following fields (the entire response should be a json and nothing else): title, description, short_summary and long_summary. {}",
        url, content
    ));
    let response = client
        .post("http://localhost:11434/api/generate")
        .json(&serde_json::json!({
            "model": model,
            "prompt": prompt,
            "stream": false,
        }))
        .send()
        .await?;

    let response_json: serde_json::Value = response.json().await?;
    let summary_text: String = response_json["response"]
        .as_str()
        .unwrap_or("No summary available")
        .to_string();
    let summary_duration: u64 = response_json["total_duration"].as_u64().unwrap_or(0);

    Ok((summary_text, summary_duration))
}

File: backend/summarizer.py
import requests
import logging

def summarize_content(url: str, model: str, content: str) -> tuple[str, int]:
    prompt = f"""Following is the text extracted from the url {url} and I want you to summarize it and give me just a crisp short summary without anything else. {content}"""

    try:
        logging.info(f"Sending summarization request for URL: {url}")

        response = requests.post(
            "http://localhost:11434/api/generate",
            json={
                "model": model,
                "prompt": prompt,
                "stream": False,
            },
        )

        response.raise_for_status()
        response_json = response.json()

        summary_text = response_json.get("response", "No summary available")
        summary_duration = int(response_json.get("total_duration", 0))

        logging.info(f"Summarization successful for URL: {url}")
        return summary_text, summary_duration

    except requests.exceptions.RequestException as e:
        logging.error(f"Error communicating with summarization API for URL {url}: {e}")
        return "Error summarizing content", 0
